Fix shadow indexing mode selection

VladLazar opened the pull request:
## Cover letter

Previously, topic level overrides of remote read/write were not honoured
in some cases. For example, if remote write was enabled at the cluster
level (cloud_storage_enable_remote_write=true) and disabled at the topic
level (redpanda.remote.write=false), remote writes would still be
performed. This is due to a bug in the logic which chooses the SI mode:
if none of the topic level overrides are true,
'get_shadow_indexing_mode` returns a std::nullopt which causes the
default value of the properties to be used. The default value is
whatever the cluster level config is set to.

This PR fixes the issue and adds ducktape tests for the SI mode selection.

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [ ] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [X] v22.2.x
- [X] v22.1.x
- [ ] v21.11.x

## UX changes
* Topic level overrides for remote read/write work in all cases now.

## Release notes
* Fix bug in remote read/write enablement. Topic level overrides are now respected in all cases.

abhijat commented on tests/rptest/tests/topic_creation_test.py:
```diff
@@ -210,6 +210,49 @@ def test_create_topic_with_single_configuration_property(self):
+    def test_shadow_indexing_mode(self, cloud_storage_enable_remote_read,
+                                  cloud_storage_enable_remote_write,
+                                  topic_remote_read, topic_remote_write):
+        to_bool = lambda x: True if x == "true" else False
```
nit: this could be simplified to:
```suggestion
        to_bool = lambda x: x == "true"
```

or maybe even compare directly:

``read = `ret["redpanda.remote.read"][0] == "true"```

jcsp commented on tests/rptest/tests/topic_creation_test.py:
```diff
@@ -210,6 +210,49 @@ def test_create_topic_with_single_configuration_property(self):
             cfgs = rpk.describe_topic_configs(topic=name)
             assert str(cfgs[p][0]) == str(property_value)
 
+    @cluster(num_nodes=3)
```
This test could probably be single node to make it a bit cheaper to run

VladLazar replied:
Done.

jcsp commented on tests/rptest/tests/topic_creation_test.py:
```diff
@@ -210,6 +210,49 @@ def test_create_topic_with_single_configuration_property(self):
             assert str(cfgs[p][0]) == str(property_value)
 
+    @cluster(num_nodes=3)
+    @matrix(cloud_storage_enable_remote_read=[True, False],
```
nit: `@matrix` is nice and expressive, but also quite expensive way to do this: it's going to start and stop a whole redpanda cluster for each permutation.

Perhaps make it into some nested for loops inside the test.

VladLazar replied:
Good idea. Done,

abhijat commented on src/v/kafka/server/handlers/topics/types.cc:
```diff
@@ -87,15 +87,27 @@ get_bool_value(const config_map_t& config, std::string_view key) {
-    auto si_enabled = get_bool_value(config, topic_property_remote_read);
-    if (si_enabled && *si_enabled) {
+    if (*si_enabled) {
         mode = mode == model::shadow_indexing_mode::archival
                  ? model::shadow_indexing_mode::full
                  : model::shadow_indexing_mode::fetch;
```
I know this is not part of this PR but I wonder if we can just have

```
mode |= model::shadow_indexing_mode::fetch;
```

here.  Maybe it is possible now, that we have a concrete value at this point rather than a possible nullopt?

VladLazar replied:
I kind of prefer the more verbose approach.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c4e997efed53f54efc2d4e818d938aadf73c78c3..55fd2bfdbe086f3576560477e9bbb0b3b6978334):
* Following a discussion with John, I realised that if the topic does not specify either remote read or write policies, it should always follow the cluster level configuration (even if it changes at a later point). I've updated the code to match this behaviour and added another ducktape test for it.
* Made the new tests faster and cheaper.

VladLazar commented:
/backport v22.2.x

VladLazar commented:
/backport v22.1.x

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the below command:

```
git cherry-pick -x 5a13e8d6936cacb1a4dea06a9b13157ab36cf7af 55fd2bfdbe086f3576560477e9bbb0b3b6978334
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/3206017935).

VladLazar commented:
/backport v22.2.x

VladLazar commented:
/backport v22.1.x

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the below command:

```
git cherry-pick -x 5a13e8d6936cacb1a4dea06a9b13157ab36cf7af 55fd2bfdbe086f3576560477e9bbb0b3b6978334
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/3221038413).

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the below command:

```
git cherry-pick -x 5a13e8d6936cacb1a4dea06a9b13157ab36cf7af 55fd2bfdbe086f3576560477e9bbb0b3b6978334
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/3221039621).
